File: Excel/479 Challenge.R
library(tidyverse)
library(readxl)
library(tictoc)
library(memoise)

path = "Excel/479 Recaman Sequence.xlsx"
test = read_excel(path)

recaman_sequence <- function(n) {
  recaman <- integer(n)
  recaman[1] <- 0
  seen <- setNames(logical(n * 3), 0:(n * 3 - 1))
  seen[1] <- TRUE
  
  for (i in 2:n) {
    prev_value <- recaman[i - 1]
    next_value <- prev_value - (i - 1)
    
    if (next_value > 0 && !seen[next_value + 1]) {
      recaman[i] <- next_value
    } else {
      next_value <- prev_value + (i - 1)
      recaman[i] <- next_value
    }
    
    seen[recaman[i] + 1] <- TRUE
  }
  
  return(recaman)
}

tic()
recaman_sequence(10000)
toc()
#  0.05 sec elapsed

result = recaman_sequence(10000)
identical(result, test$`Answer Expected`)
# [1] TRUE

File: Excel/479 Challenge.py
import pandas as pd
import time

path = "479 Recaman Sequence.xlsx"
test = pd.read_excel(path)

def recaman_sequence(n):
    recaman = [0] * n
    seen = set()
    seen.add(0)

    for i in range(1, n):
        next_value = recaman[i - 1] - i
        if next_value > 0 and next_value not in seen:
            recaman[i] = next_value
        else:
            next_value = recaman[i - 1] + i
            recaman[i] = next_value
        seen.add(recaman[i])

    return recaman

start_time = time.time()
sequence = recaman_sequence(10000)
end_time = time.time()
print(f"Time taken: {end_time - start_time} seconds")
# Time taken: 0.0019996166229248047 seconds

print(all(sequence[i] == test["Answer Expected"][i] for i in range(len(sequence))))
# True